### PR TITLE
Add ability to run specific tests with Secure Boot

### PR DESCRIPTION
**Pull request by Jedoku**

### CODE REVIEW

**Review by akihikodaki — CHANGES_REQUESTED** [review]
Don't add `stop`. Instead, open a `ResourceScope` and run clients in the scope for each test group.

File: lib/engines/hcktest/hcktest.rb [hunk 1]
```diff
@@ -151,21 +151,28 @@ def configure_clients
       end
     end
 
+    def reconfigure_clients
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:154]: This name does not represent what it actually does. "Reconfiguring" sounds like it is a special operation that corresponds to `configure_clients` but should be used for a second time. But it does more than configure (it also synchronizes). And such a sequence of configuring and synchronizing is also valid for the first run.

File: lib/engines/hcktest/hcktest.rb [hunk 2]
```diff
@@ -151,21 +151,28 @@ def configure_clients
...
+      configure_clients
+      @clients.each_value(&:synchronize)
+    end
+
     def configure_setup_and_synchronize
+      return reconfigure_clients unless @studio.tools.nil?
+
       @studio.configure(@platform['clients'])
       configure_clients
       @clients.each_value(&:synchronize)
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:164]: Call `reconfigure_clients` here (and `reconfigure_clients` should be renamed as it is awkward to call like that when it is used for the first run).

File: lib/engines/hcktest/hcktest.rb [hunk 3]
```diff
@@ -151,21 +151,28 @@ def configure_clients
...
+    def run_and_configure_clients(scope)
       retries = 0
       begin
         scope.transaction do |tmp_scope|
-          run_studio tmp_scope
-          sleep 5 until @studio.up?
-          run_clients tmp_scope, keep_alive: true
+          run_opts = { keep_alive: true }
+          run_opts.merge!(@client_run_opts) if @client_run_opts
+          run_clients tmp_scope, run_opts
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:175]:
```suggestion
          run_clients tmp_scope, keep_alive: true, **@client_run_opts
```

File: lib/engines/hcktest/hcktest.rb [hunk 4]
```diff
@@ -151,21 +151,28 @@ def configure_clients
...
+
       @studio.configure(@platform['clients'])
       configure_clients
       @clients.each_value(&:synchronize)
       @studio.keep_snapshot
       @clients.each_value(&:keep_snapshot)
     end
 
-    def run_and_configure_setup(scope)
+    def run_and_configure_clients(scope)
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:169]: Drop `setup` everywhere if you are dropping it here.

File: lib/engines/hcktest/hcktest.rb (hunk 3 above)
Comment (akihikodaki) [reply lib/engines/hcktest/hcktest.rb:175]: I also want you to minimize the use of instance variables; pass `@client_run_opts` as an argument of this method instead.

File: lib/engines/hcktest/hcktest.rb [hunk 5]
```diff
@@ -225,22 +232,69 @@ def prepare_tests
...
+
+        prepare_environment(group)
+        ResourceScope.open do |scope|
+          run_and_configure_clients scope
+
+          @tests.run(tests)
+          break if @project.run_terminated
+        end
+      end
+    end
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:265]: This is a bit strange division of concerns. There are two things in `run_tests`; running "normal" tests and running the other tests. However, while the first is abstracted away, the second isn't.

We can extract the second away by creating another method. Not being "normal" is "abnormal" but that sounds bad in English. Perhaps you may name it e.g., `run_tests_without_config` and `run_tests_with_config` to represent what it does more precisely; it is verbose but I think that's fine since it's only called from one place anyway.

File: lib/engines/hcktest/hcktest.rb [hunk 6]
```diff
@@ -225,22 +232,69 @@ def prepare_tests
...
+        sleep 5 until @studio.up?
 
-        @tests.run
+        run_tests
 
         pause_run if @project.options.test.manual
 
         @tests.create_project_package
       end
     end
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:249]: What should be done for `dump_run`?

**Review by akihikodaki — CHANGES_REQUESTED** [review]

File: lib/engines/hcktest/hcktest.rb [hunk 7]
```diff
@@ -225,19 +226,64 @@ def prepare_tests
...
+      grouped_tests = { normal: [], secure: [] }
+
+      @config.tests_config.each do |test_group|
+        selected_tests = @test_list.select { |test| test_group.tests.include?(test['name']) }
+        grouped_tests[:secure] += selected_tests if test_group.secure
+      end
+
+      grouped_tests[:normal] += (@test_list - grouped_tests.values.flatten)
+      grouped_tests
+    end
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:280]: Let's split this into a method that returns tests with config and another that returns tests without config.

File: lib/engines/hcktest/hcktest.rb [hunk 8]
```diff
@@ -143,29 +143,30 @@ def run_clients(scope, run_opts = {})
...
       @studio.configure(@platform['clients'])
-      configure_clients
-      @clients.each_value(&:synchronize)
+      configure_and_synchronize_clients
       @studio.keep_snapshot
       @clients.each_value(&:keep_snapshot)
     end
 
-    def run_and_configure_setup(scope)
+    def run_clients_and_configure_setup(scope, client_run_opts = {})
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:165]: Simpler:

```suggestion
    def run_clients_and_configure_setup(scope, **opts)
```

We already say `run_clients` so we don't need to say that in this method. Also let's use `**` trick.

File: lib/engines/hcktest/hcktest.rb [hunk 9]
```diff
@@ -225,19 +226,64 @@ def prepare_tests
...
+      end
+    end
 
-        @tests.create_project_package
+    def run_tests_with_config
+      group_tests.each do |group, tests|
+        next if tests.empty? || group == :normal
+
+        ResourceScope.open do |scope|
+          run_clients_and_configure_setup(scope, client_run_options(group))
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:260]: I guess you can just do:

```suggestion
          run_clients_and_configure_setup(scope, group => true)
```

File: lib/engines/hcktest/hcktest.rb [hunk 10]
```diff
@@ -225,19 +226,64 @@ def prepare_tests
...
+
+          @tests.run(tests)
+
+          pause_run if @project.options.test.manual
+          break if @project.run_terminated
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:265]: Why do we need this line though we didn't have one previously and don't have one even now for the tests without config?

**Review by akihikodaki — CHANGES_REQUESTED** [review]
I found the logging code is inconsistent. For the tests without config, it logs after configuring the clients. For tests with configs, it logs before configuring clients.

Let's also not introduce a new term "environment" but an existing word "clients".

File: lib/setupmanagers/hckclient.rb [hunk 11]
```diff
@@ -134,6 +149,8 @@ def initialize_client_wait
     end
 
     def return_when_client_up
+      move_machine_to_default_pool if machine_in_working_pool?
```
Comment (kostyanf14) [inline lib/setupmanagers/hckclient.rb:152]: Why do we need this? Let's just skip recognize_client_wait and initialize_client_wait if client initialized and not in Default Pool.

File: lib/engines/hcktest/hcktest.rb [hunk 12]
```diff
@@ -143,29 +143,30 @@ def run_clients(scope, run_opts = {})
...
 
-    def run_and_configure_setup(scope)
+    def run_clients_and_configure_setup(scope, client_run_opts = {})
       retries = 0
       begin
         scope.transaction do |tmp_scope|
-          run_studio tmp_scope
-          sleep 5 until @studio.up?
-          run_clients tmp_scope, keep_alive: true
+          run_clients tmp_scope, keep_alive: true, **client_run_opts
```
Comment (kostyanf14) [inline lib/engines/hcktest/hcktest.rb:169]: I see a problem with the current implementation of HCKClient recreation. When we create a new HCKClient it will call setup_manager to create a snapshot again. In this case, we lose information from the previous run (can capture files from a real VM if needed). 

There are 2 ways:
- add a new option to reuse existing snapshot
- add a new option to rename snapshots according to the test group

@akihikodaki What do you think?

**Review by kostyanf14 — CHANGES_REQUESTED** [review]

File: lib/engines/hcktest/hcktest.rb (hunk 12 above)
Comment (akihikodaki) [reply lib/engines/hcktest/hcktest.rb:169]: I'm for the former. We don't have a reason to create a snapshot for each test group while we don't create one for each test.

File: lib/engines/hcktest/hcktest.rb [hunk 13]
```diff
@@ -221,20 +222,56 @@ def prepare_tests
...
+        @logger.info('Environment ready, running normal tests')
+        @tests.run(@test_list - group_tests_by_config.values.flatten)
         pause_run if @project.options.test.manual
+      end
+    end
 
-        @tests.create_project_package
+    def run_tests_with_config
+      group_tests_by_config.each do |group, tests|
+        next if tests.empty? || group == :normal
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:253]:
```suggestion
        next if tests.empty?
```

File: lib/setupmanagers/hckclient.rb [hunk 14]
```diff
@@ -55,7 +55,13 @@ def restart_machine
       @tools.restart_machine(@name)
     end
 
+    def reconfigure_pool
+      @pool = @project.engine_tag
+    end
```
Comment (akihikodaki) [inline lib/setupmanagers/hckclient.rb:60]: Having `@pool = 'Default Pool'` does no longer make sense in the first place. Let's remove `@pool` and instead create a method `pool` that queries and returns the pool.

**Review by akihikodaki — CHANGES_REQUESTED** [review]

File: lib/setupmanagers/hckclient.rb [hunk 15]
```diff
@@ -46,8 +49,8 @@ def reconfigure_machine
     end
 
     def delete_machine
-      @tools.delete_machine(@name, @pool)
-      @pool = 'Default Pool'
+      @tools.delete_machine(@name, pool)
+      pool
```
Comment (akihikodaki) [inline lib/setupmanagers/hckclient.rb:53]: I don't think we need to call `pool` here.

File: lib/setupmanagers/hckclient.rb [hunk 16]
```diff
@@ -18,23 +18,26 @@ def initialize(setup_manager, scope, studio, name, run_opts)
...
       @name = name
       @kit = setup_manager.kit
-      @pool = 'Default Pool'
       @logger.info("Starting client #{name}")
       @runner = setup_manager.run_client(scope, @name, run_opts)
       scope << self
     end
 
+    def pool
+      machine_in_working_pool? ? @project.engine_tag : 'Default Pool'
```
Comment (akihikodaki) [inline lib/setupmanagers/hckclient.rb:27]: You can instead use `@studio.list_pools` and return the pool containing this machine.

File: lib/setupmanagers/hckclient.rb [hunk 17]
```diff
@@ -139,21 +149,27 @@ def return_when_client_up
...
+          prepare_envirenment
         end
 
-        @logger.info("Preparing client #{@name}...")
-        @project.extra_sw_manager.install_software_before_driver(@tools, @name)
-        install_drivers
-        copy_dvl
-        @project.extra_sw_manager.install_software_after_driver(@tools, @name)
         @logger.info("Configuring client #{@name}...")
         configure_machine
```
Comment (akihikodaki) [inline lib/setupmanagers/hckclient.rb:174]: Don't call `move_machine_to_pool` when `machine_in_working_pool?`.

**Review by akihikodaki — CHANGES_REQUESTED** [review]

File: lib/setupmanagers/hckclient.rb [hunk 18]
```diff
@@ -18,36 +18,42 @@ def initialize(setup_manager, scope, studio, name, run_opts)
...
       @runner = setup_manager.run_client(scope, @name, run_opts)
       scope << self
     end
 
+    def pool
+      @studio.list_pools.each do |pool|
+        return pool['name'] if pool['machines'].any? { |machine| machine['name'].eql?(@name) }
+      end
+
+      'Default Pool'
```
Comment (akihikodaki) [inline lib/setupmanagers/hckclient.rb:31]: It shouldn't return `'Default Pool'` when `list_pools` says it's not in `Default Pool`.

File: lib/setupmanagers/hckclient.rb [hunk 19]
```diff
@@ -114,9 +119,16 @@ def copy_dvl
       end
     end
 
+    def machine_in_working_pool?
```
Comment (akihikodaki) [inline lib/setupmanagers/hckclient.rb:122]: This method should no longer be necessary. You can just call `pool` and check the returned value.

File: lib/setupmanagers/hckclient.rb [hunk 20]
```diff
@@ -139,21 +151,28 @@ def return_when_client_up
       initialize_client_wait
     end
 
+    def prepare_envirenment
```
Comment (akihikodaki) [inline lib/setupmanagers/hckclient.rb:154]: "Environment" is not a word used elsewhere. Name it `prepare_client` or `prepare_machine`.

File: lib/setupmanagers/hckclient.rb [hunk 21]
```diff
@@ -169,7 +188,7 @@ def synchronize
     end
 
     def not_ready?
-      @studio.list_pools.detect { |pool| pool['name'].eql?(@pool) }['machines']
+      @studio.list_pools.detect { |p| p['name'].eql?(pool) }['machines']
```
Comment (akihikodaki) [inline lib/setupmanagers/hckclient.rb:191]: This call of `pool` is unnecessary.

**Review by akihikodaki — CHANGES_REQUESTED** [review]

**Comment by kostyanf14** [comment]
@Jedoku HCK-CI run failed

```
00:02:40.087 F, [2024-08-28T02:50:10.086695 #148260] FATAL -- : (NoMethodError) undefined method `[]' for nil:NilClass
00:02:40.087    -- AutoHCK-PR-run/lib/setupmanagers/hckclient.rb:182:in `not_ready?'
00:02:40.087    -- AutoHCK-PR-run/lib/setupmanagers/hckclient.rb:188:in `reset_to_ready_state'
00:02:40.087    -- AutoHCK-PR-run/lib/setupmanagers/hckclient.rb:41:in `configure_machine'
00:02:40.087    -- AutoHCK-PR-run/lib/setupmanagers/hckclient.rb:168:in `block in configure'
00:02:40.087    -- AutoHCK-PR-run/vendor/bundle/ruby/3.2.0/gems/logging-2.4.0/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
```

File: lib/setupmanagers/hckclient.rb [hunk 22]
```diff
@@ -116,7 +119,7 @@ def copy_dvl
 
     def machine_in_default_pool
```
Comment (akihikodaki) [inline lib/setupmanagers/hckclient.rb:120]: This method is no longer necessary; it can be simply replaced with `pool == 'Default Pool'`

File: lib/setupmanagers/hckclient.rb [hunk 23]
```diff
@@ -139,21 +142,28 @@ def return_when_client_up
       initialize_client_wait
     end
 
+    def prepare_machine
+      @logger.info("Preparing client #{@name}...")
```
Comment (akihikodaki) [inline lib/setupmanagers/hckclient.rb:146]: `configure_machine` doesn't contain `@logger.info("Configuring client #{@name}...")` but `prepare_machine` does. Keep them consistent.

File: lib/engines/hcktest/hcktest.rb [hunk 24]
```diff
@@ -229,20 +230,56 @@ def prepare_tests
...
         pause_run if @project.options.test.manual
+      end
+    end
 
-        @tests.create_project_package
+    def run_tests_with_config
+      group_tests_by_config.each do |group, tests|
+        next if tests.empty?
+
+        @logger.info("Preparing environment for #{group} tests")
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:263]: Log just before running tests instead of logging here because:
- `HCKClient#configure` logs before configuring anyway.
- It is inconsistent with `run_tests_without_config`, which logs just before running tests.

**Review by akihikodaki — CHANGES_REQUESTED** [review]

**Review by akihikodaki — CHANGES_REQUESTED** [review]
Please check the unresolved comments.

File: lib/engines/hcktest/hcktest.rb (hunk 6 above)
Comment (akihikodaki) [reply lib/engines/hcktest/hcktest.rb]: Please add code to raise an error for `dump_run`

File: lib/engines/hcktest/hcktest.rb [hunk 25]
```diff
@@ -254,6 +292,9 @@ def run
         dump_run
 
         @project.logger.info("Find all scripts in folder: #{@project.workspace_path}")
+        @project.logger.warn(
+          'Dump mode is only for basic tests. For specific configurations (e.g., Secure Boot tests), do it manually.'
```
Comment (kostyanf14) [inline lib/engines/hcktest/hcktest.rb:296]: do it manually -> update run scripts manually

**Review by kostyanf14 — APPROVED** [review]

File: lib/engines/hcktest/hcktest.rb [hunk 26]
```diff
@@ -254,6 +292,10 @@ def run
         dump_run
 
         @project.logger.info("Find all scripts in folder: #{@project.workspace_path}")
+        @project.logger.warn(
+          'Dump mode is only for basic tests. ' \
+          'For specific configurations (e.g., Secure Boot tests), update run scripts manually'
+        )
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:298]: Split this into two `warn` calls; if the message is too long in source code, it is probably too long to fit in the terminal as well.

File: lib/engines/hcktest/tests.rb [hunk 27]
```diff
@@ -6,7 +6,7 @@ module AutoHCK
   class Tests
     include Helper
 
-    HANDLE_TESTS_POLLING_INTERVAL = 10
+    HANDLE_TESTS_POLLING_INTERVAL = 30
```
Comment (akihikodaki) [inline lib/engines/hcktest/tests.rb:9]: Is this change really necessary?

**Review by akihikodaki — CHANGES_REQUESTED** [review]

File: lib/engines/hcktest/hcktest.rb [hunk 28]
```diff
@@ -254,6 +292,10 @@ def run
         dump_run
 
         @project.logger.info("Find all scripts in folder: #{@project.workspace_path}")
+        @project.logger.warn(
+          'Dump mode is only for basic tests. ' \
+          'For specific configurations (e.g., Secure Boot tests), update run scripts manually'
```
Comment (akihikodaki) [inline lib/engines/hcktest/hcktest.rb:297]: "run" is superfluous; I don't remember any code that calls the scripts "run scripts". You can just write as "scripts".

**Review by akihikodaki — APPROVED** [review]
